### ISSUE TITLE
Tight coupling of ja2-straciatella sources [meta]

### ISSUE THREAD

**Issue by janisozaur** [ja2-stracciatella/ja2-stracciatella#454]
Using a tool: https://github.com/tomtom-international/cpp-dependencies I created a graph of dependencies in ja2-straciatella (as of 9da7968).

![graph](https://cloud.githubusercontent.com/assets/550290/20944389/86b28454-bc03-11e6-98b6-3f83826be172.png)

What you can see in this graph (having familiarised yourself with the tool in question) is that most dependencies within ja2 sources are coloured orange, which indicate cycles. Such cycles don't yield easily to testing, especially when trying to build small units of code to test.

It's nice to see any tests in such a project really, but perhaps you could consider improving architecture of these components so they can be used and tested more easily?

**Comment by Manuel-K** [comment]
I agree. Some refactoring could improve the codebase, but I think #360 should be implemented first because the current structure is confusing.

**Comment by lynxlynxlynx** [comment]
and to think that this is the state after years of cleanups and refactoring already. All those horrors we have been saved from. :)

**Comment by selaux** [comment]
I agree. This also makes the learning curve quite steep for new developers to enter. On the other hand it is really hard to get rid of this. The amount of global stuff that is used throughout the codebase is just too damn high. This was actually one of the reasons I started the Rust stuff, it forces you to some degree to get rid of the global dependencies of functions (also because you cannot call C++ from Rust). Time will tell if that is actually true 😉.

**Comment by lynxlynxlynx** [comment]
On a related note, just ripping the gui stuff out into a separate (data) format would probably get rid of a few 10% of preprocessor defines. With a bit more wm work, it would also clean up the setup and interactivity, improving readability further. No real gain though, until someone wants to do better resolution stuff.

**Comment by janisozaur** [comment]
I saw the bits in Rust were recently merged, this looks like a really nice progress 👍 . Rust also looks like a very interesting language, I'd love to have time to contribute to Rust-ification of ja2-straciatella.

**Comment by derSteFfi** [comment]
isolation the gui data would help a lot. we could store it in one place and it would help for the resolution stuff. Also, there once was the idea of smooth scrolling and it would be easier with isolated gui data

**Comment by Manuel-K** [comment]
Version of the graph with reorganized sources (#360):

![graph](https://cloud.githubusercontent.com/assets/775439/21113667/baf6c09e-c0aa-11e6-8e7a-6b62ee53880d.png)

**Comment by janisozaur** [comment]
@Manuel-K can you post output from `--stats` from "before" and "after"?

**Comment by Manuel-K** [comment]
No changes so far. I've basically renamed folders.

Before:
```
git checkout master
 ../cpp-dependencies.bin --stats
20 components with 43 public dependencies, 75 private dependencies
Detected 13 nodes in cycles
```

After:
```
git checkout folder_renames
../cpp-dependencies.bin --stats
20 components with 43 public dependencies, 75 private dependencies
Detected 13 nodes in cycles
```

I'm looking for a way to untangle at least some of the cycles… it may take some time.

**Comment by selaux** [comment]
Does it check for usage? Because my IDE (CLion) tells me a lot of headers are unused...

**Comment by janisozaur** [comment]
@selaux no, it doesn't really understand the language; it just parses the `#include`s.

I'd also suggest using a proper tool like IWYU over CLion's one.

**Comment by Manuel-K** [comment]
Somewhat. It does check for unused files. There are many of them, especially parts of boost. As @janisozaur  said, it is very limited. I assume that there are some false positives.

Output of `../cpp-dependencies.bin --outliers`: [outliers.txt](https://github.com/ja2-stracciatella/ja2-stracciatella/files/647139/outliers.txt)

**Comment by Manuel-K** [comment]
Moving the all unittests to subfolders of src/unittests does clean up the graph somewhat, but that would only be a cosmetic change for now:
![graph3](https://cloud.githubusercontent.com/assets/775439/21116306/74088914-c0b5-11e6-855e-72b4ddf574a6.png)

```
../cpp-dependencies.bin --stats
26 components with 44 public dependencies, 93 private dependencies
Detected 13 nodes in cycles
```
